pkg/policy/rule.go

```diff
@@ -41,14 +41,16 @@ func (r *rule) validate() error {
 	return nil
 }
 
-func mergeL4Port(ctx *SearchContext, r api.PortRule, p api.PortProtocol, dir string, proto string, resMap L4PolicyMap) int {
+func mergeL4Port(fromEndpoints []api.EndpointSelector, r api.PortRule, p api.PortProtocol,
+	dir string, proto string, resMap L4PolicyMap) int {
+
 	fmt := p.Port + "/" + proto
 	v, ok := resMap[fmt]
 	if !ok {
-		resMap[fmt] = CreateL4Filter(r, p, dir, proto)
+		resMap[fmt] = CreateL4Filter(fromEndpoints, r, p, dir, proto)
 		return 1
 	}
-	l4Filter := CreateL4Filter(r, p, dir, proto)
+	l4Filter := CreateL4Filter(fromEndpoints, r, p, dir, proto)
```

Comment:
This may need further thought. Imagine the case where we have the following rules:
```
  ingress:
  - fromEndpoints:
    - matchLabels:
        id: app2
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP
      rules:
        http:
        - method: "GET"
          path: "/public"
  - toPorts:
    - ports:
      - port: "80"
        protocol: TCP
      rules:
        http:
        - method: "GET"
          path: "/public"
```

I think we would agree that if we have two rules which both cover port 80/tcp. One is restricted to fromEndpoints, one is not. The one without fromEndpoints should take precedence as it also includes the more restrictive rule.

I thik we need to clear existing fromEndpoints conditions if we encounter a unrestricted toPorts and we need to skip conditional toPorts if we already have an unrestricted L4 rule.

L7 rule will potentially need to be stored per fromEndpoints as the rules should not apply to other endpoints. So while we want to potentially skip/clear fromEndpoints restrictions on pure L4 allow/disallow. We always need to store the L7 rules per fromEndpoints. We will need to encode logic in the proxy which allows matching on the source identity.

What do you think?